File: packages/shared/src/ticket.ts
/**
 * Contratti condivisi tra desktop e mobile.
 * Devono restare allineati con gli enum del backend (backend/src/app/models/ticket.py).
 */

/** Stati del ciclo di vita di un ticket. */
export enum TicketStatus {
  Creato = 'creato',
  InLavorazione = 'in_lavorazione',
  InAttesa = 'in_attesa',
  Approvato = 'approvato',
  Concluso = 'concluso',
  Rifiutato = 'rifiutato',
}

/** Tipo di ticket: determina come l'AI lo lavora. */
export enum TicketType {
  Email = 'email',
  Fix = 'fix',
  Feature = 'feature',
}

/** Origine del ticket. */
export enum TicketSource {
  Manuale = 'manuale',
  Email = 'email',
  Odoo = 'odoo',
}

/** Ordine di visualizzazione delle colonne nella board. */
export const STATUS_ORDER: TicketStatus[] = [
  TicketStatus.Creato,
  TicketStatus.InLavorazione,
  TicketStatus.InAttesa,
  TicketStatus.Approvato,
  TicketStatus.Concluso,
];

/** Etichette leggibili (IT) per la UI. */
export const STATUS_LABELS: Record<TicketStatus, string> = {
  [TicketStatus.Creato]: 'Creato',
  [TicketStatus.InLavorazione]: 'In lavorazione',
  [TicketStatus.InAttesa]: 'In attesa',
  [TicketStatus.Approvato]: 'Approvato',
  [TicketStatus.Concluso]: 'Concluso',
  [TicketStatus.Rifiutato]: 'Rifiutato',
};

export const TYPE_LABELS: Record<TicketType, string> = {
  [TicketType.Email]: 'Email',
  [TicketType.Fix]: 'Fix',
  [TicketType.Feature]: 'Feature',
};

export interface Ticket {
  id: number;
  title: string;
  description: string | null;
  type: TicketType;
  status: TicketStatus;
  source: TicketSource;
  /** Nota dell'AI mostrata all'utente (es. "va approvata l'email"). */
  ai_note: string | null;
  /** Output dell'AI da rivedere: bozza email o piano di modifica. */
  ai_draft: string | null;
  /** Nota dell'utente in caso di rifiuto/revisione. */
  review_note: string | null;
  branch_name: string | null;
  external_ref: string | null;
  /** Per i ticket email: indirizzo a cui rispondere e account di provenienza. */
  source_address: string | null;
  email_account_id: number | null;
  /** Per i ticket fix/feature: progetto git su cui lavorare. */
  project_id: number | null;
  created_at: string;
  updated_at: string;
}

/** Esito di un giro del job worker. */
export interface JobRunResult {
  processed: number;
  finalized: number;
  results: { ticket_id: number; action: string; status: string; note: string }[];
  errors: string[];
}

/** Payload per creare un ticket. */
export interface CreateTicketInput {
  title: string;
  description?: string | null;
  type: TicketType;
  source?: TicketSource;
  external_ref?: string | null;
  project_id?: number | null;
}

/** Payload per aggiornare lo stato (con eventuale nota di revisione). */
export interface UpdateTicketStatusInput {
  status: TicketStatus;
  review_note?: string | null;
}

File: backend/src/app/models/ticket.py
"""Model ORM del ticket.

Gli enum devono restare allineati con i contratti TS condivisi
(packages/shared/src/ticket.ts).
"""

import enum
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class TicketStatus(str, enum.Enum):
    CREATO = "creato"
    IN_LAVORAZIONE = "in_lavorazione"
    IN_ATTESA = "in_attesa"
    APPROVATO = "approvato"
    CONCLUSO = "concluso"
    RIFIUTATO = "rifiutato"


class TicketType(str, enum.Enum):
    EMAIL = "email"
    FIX = "fix"
    FEATURE = "feature"


class TicketSource(str, enum.Enum):
    MANUALE = "manuale"
    EMAIL = "email"
    ODOO = "odoo"


class Ticket(Base):
    __tablename__ = "tickets"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    type: Mapped[TicketType] = mapped_column(Enum(TicketType), nullable=False)
    status: Mapped[TicketStatus] = mapped_column(
        Enum(TicketStatus), nullable=False, default=TicketStatus.CREATO, index=True
    )
    source: Mapped[TicketSource] = mapped_column(
        Enum(TicketSource), nullable=False, default=TicketSource.MANUALE
    )

    # Nota prodotta dall'AI (es. "va approvata l'email"); compilata allo Step 3.
    ai_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    # Output dell'AI da rivedere/approvare: bozza email o piano di modifica codice.
    ai_draft: Mapped[str | None] = mapped_column(Text, nullable=True)
    # Nota dell'utente in caso di rifiuto/revisione.
    review_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    # Branch git per ticket di tipo fix/feature (Step 4).
    branch_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    # Riferimento esterno (id email/thread o id ticket Odoo) (Step 2/5).
    external_ref: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # --- Provenienza email (per ticket type=email): a chi rispondere e da quale account ---
    source_address: Mapped[str | None] = mapped_column(String(320), nullable=True)
    email_account_id: Mapped[int | None] = mapped_column(
        ForeignKey("email_accounts.id", ondelete="SET NULL"), nullable=True
    )

    # --- Progetto git (per ticket type=fix/feature) ---
    project_id: Mapped[int | None] = mapped_column(
        ForeignKey("projects.id", ondelete="SET NULL"), nullable=True
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now(), nullable=False
    )
